src/coding-guidelines/types-and-traits.rst

```diff
@@ -63,3 +64,225 @@ Types and Traits
              current.checked_add(velocity).expect("Position calculation overflowed")
          }
 
+.. guideline:: Do Not Depend on Function Pointer Identity Across Crates
+    :id: gui_QbvIknd9qNF6 
+    :category: required
+    :status: draft
+    :release: unclear-latest
+    :fls: fls_1kg1mknf4yx7
+    :decidability: decidable
+    :scope: system
+    :tags: surprising-behavior
+
+    Do not rely on the equality or stable identity of function pointers originating from different crates or that may be inlined,
+    duplicated, or instantiated differently across compilation units, codegen units, or optimization profiles.
+
+    Avoid assumptions about low-level metadata (such as symbol addresses) unless explicitly guaranteed by the Ferrocene Language Specification (FLS).
+    Function address identity is not guaranteed by Rust and must not be treated as stable.
+    Rust’s ``fn`` type is a zero-sized function item promoted to a function pointer, whose address is determined by the compiler backend.
+    When a function resides in a different crate, or when optimizations such as inlining,
+    link-time optimization, or codegen-unit partitioning are enabled,
+    the compiler may generate multiple distinct code instances for the same function or alter the address at which it is emitted.
+
+    Consequently, the following operations are not reliable:
+
+    - Comparing function pointers for equality (``fn1 == fn2``)
+    - Assuming a unique function address
+    - Using function pointers as identity keys (e.g., in maps, registries, matchers)
+    - Matching behavior based on function address
+
+    This rule applies even when the functions are semantically identical, exported as ``pub``, or defined once in source form.
+
+    .. rationale:: 
+        :id: rat_xcVE5Hfnbb2u 
+        :status: draft
+
+        Compiler optimizations may cause function pointers originating from different crates to lose stable identity.
+        Observed behaviors include:
+
+        - Cross-crate inlining producing multiple code instantiations
+        - Codegen-unit separation causing function emission in multiple units
+        - Incremental builds producing variant symbol addresses
+        - Link-time optimization merging or splitting functions unpredictably
```

Comment:
Nit: functions can be merged even when LTO is not enabled

Reply:
resolved